store/daystore/day_store.go

```diff
@@ -0,0 +1,212 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package store
+
+import (
+	"fmt"
+	"math"
+	"sort"
+	"sync"
+	"time"
+
+	ss "k8s.io/heapster/store/statstore"
+
+	window "k8s.io/heapster/third_party/window"
+)
+
+// DayStore is an in-memory window of derived stats for each hour of a day.
+// DayStore holds 24 hours of derived stats, plus an hour-long StatStore that holds -
+// historical data on the current hour.
+// DayStore can calculate the Average, Max and 95th Percentile over the past 24 hours.
+// The latest one hour is only considered for the calculation of the daily Max.
+type DayStore struct {
+	// a RWMutex guards all operations on the underlying window and cached values.
+	sync.RWMutex
+
+	// Hour is a StatStore with data from the last one hour
+	Hour *ss.StatStore
+
+	window *window.MovingWindow
+
+	// size is the number of items currently stored in the window
+	size int
+
+	// lastFlush is the time when the previous hourly stats were flushed into the DayStore
+	lastFlush time.Time
+
+	// validAvgPct marks whether the cached average and percentiles are correct.
+	validAvgPct bool
+	// validMax marks whether the cached max value is correct.
+	validMax bool
+
+	cachedAverage     uint64
+	cachedMax         uint64
+	cachedNinetyFifth uint64
+}
+
+// hourEntry is the set of derived stats that are maintained per hour.
+type hourEntry struct {
+	average     uint64
+	max         uint64
+	ninetyFifth uint64
+}
+
+// NewDayStore is a DayStore constructor.
+// The recommended minimum resolution is at least one minute.
+func NewDayStore(epsilon uint64, resolution time.Duration) *DayStore {
+	// Calculate how many resolutions correspond to an hour
+	hourNS := time.Hour.Nanoseconds()
+	resNS := resolution.Nanoseconds()
+	intervals := uint8(hourNS / resNS)
+	if hourNS%resNS != 0 {
+		intervals++
+	}
+
+	return &DayStore{
+		window: window.New(24, 1),
+		Hour:   ss.NewStatStore(epsilon, resolution, intervals, []float64{0.95}),
+	}
+}
+
+// Put stores a TimePoint into the Hour StatStore, while checking whether it -
+// is time to flush the last hour's stats in the window.
+func (ds *DayStore) Put(tp ss.TimePoint) error {
+	ds.Lock()
+	defer ds.Unlock()
+
+	err := ds.Hour.Put(tp)
+	if err != nil {
+		return err
+	}
+
+	ds.validMax = false
+
+	// Check if this is the first TimePoint ever, in which case flush in one hour.
+	if ds.lastFlush.Equal(time.Time{}) {
+		ds.lastFlush = tp.Timestamp
+		return nil
+	}
+
+	// The new TimePoint is newer by at least one hour since the last flush
+	// flush the current hour to the window.
+	if !tp.Timestamp.Add(-time.Hour).Before(ds.lastFlush) {
```

Comment:
Looks like this is the normal flow of the code. How about inverting the `if`, making it return and then moving the rest of the code out of the block?